Always include autograd context id in rpc/remote requests

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29543 [WIP] Add end-to-end test for RNN Module
* **#29781 Always include autograd context id in rpc/remote requests**
* #29634 Allow rpc.remote to create RRef on self

Even though the request might not contain any requires_grad tensor,
the return value could. Therefore, we should always include the
autograd context id in the request.

closes #28819

Differential Revision: [D18496709](https://our.internmc.facebook.com/intern/diff/D18496709)

mrshenli committed:
Always include autograd context id in rpc/remote requests

Even though the request might not contain any requires_grad tensor,
the return value could. Therefore, we should always include the
autograd context id in the request.

[ghstack-poisoned]

mrshenli committed:
Update on "Always include autograd context id in rpc/remote requests"


Even though the request might not contain any requires_grad tensor,
the return value could. Therefore, we should always include the
autograd context id in the request.

closes #28819 

[ghstack-poisoned]

mrshenli committed:
Update on "Always include autograd context id in rpc/remote requests"


Even though the request might not contain any requires_grad tensor,
the return value could. Therefore, we should always include the
autograd context id in the request.

closes #28819

Differential Revision: [D18496709](https://our.internmc.facebook.com/intern/diff/D18496709)

[ghstack-poisoned]

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -599,6 +600,42 @@ def test_no_graph_with_tensors_not_require_grad(self):
+            send_functions = ctx._send_functions()
+            self.assertEqual(len(send_functions), 0)
+            recv_functions = ctx._recv_functions()
+            self.assertEqual(len(recv_functions), 1)
```
Instead of checking the autograd functions, can we run the backward pass and verify the gradient is accumulated correctly on the remote tensor? We probably might need to return a global tensor from `ret_requires_grad` to validate the grads are accumulated for `ExecMode.RPC_SYNC`

mrshenli replied:
Yes, let me add that.

pritamdamania87 reviewed (changes requested):
Thanks for fixing this! Just have one comment inline for the unit test.

mrshenli committed:
Update on "Always include autograd context id in rpc/remote requests"


Even though the request might not contain any requires_grad tensor,
the return value could. Therefore, we should always include the
autograd context id in the request.

closes #28819

Differential Revision: [D18496709](https://our.internmc.facebook.com/intern/diff/D18496709)

[ghstack-poisoned]

mrshenli committed:
Update on "Always include autograd context id in rpc/remote requests"


Even though the request might not contain any requires_grad tensor,
the return value could. Therefore, we should always include the
autograd context id in the request.

closes #28819

Differential Revision: [D18496709](https://our.internmc.facebook.com/intern/diff/D18496709)

[ghstack-poisoned]

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -599,6 +602,41 @@ def test_no_graph_with_tensors_not_require_grad(self):
+
+            # Wait for the prev rank to be done with rpc.
+            self._check_rpc_done(1)
+            grads = dist_autograd.get_gradients(ctx_ids[1])
```
nit: can we also check that the size of `grads` is 1?

mrshenli replied:
yes, let me add.

mrshenli committed:
Update on "Always include autograd context id in rpc/remote requests"


Even though the request might not contain any requires_grad tensor,
the return value could. Therefore, we should always include the
autograd context id in the request.

closes #28819

Differential Revision: [D18496709](https://our.internmc.facebook.com/intern/diff/D18496709)

[ghstack-poisoned]

mrshenli commented:
The lint failure is on a file that I didn't touch:

```
/home/vsts/work/1/s/aten/src/ATen/cuda/ATenCUDAGeneral.h:3:10: error: 'cuda.h' file not found [clang-diagnostic-error]
```
